core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,334 @@
+        val updatedLeaderIsrAndControllerEpochs = removeReplicasFromIsr(replicaId, replicasToRemoveFromIsr.map(_.topicAndPartition))
+        updatedLeaderIsrAndControllerEpochs.foreach { case (partition, leaderIsrAndControllerEpoch) =>
+          if (!controller.topicDeletionManager.isPartitionToBeDeleted(partition)) {
+            val recipients = controllerContext.partitionReplicaAssignment(partition).filterNot(_ == replicaId)
+            controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipients,
+              partition.topic,
+              partition.partition,
+              leaderIsrAndControllerEpoch,
+              controllerContext.partitionReplicaAssignment(partition))
+          }
+          replicaState.put(PartitionAndReplica(partition.topic, partition.partition, replicaId), OfflineReplica)
+        }
+      case ReplicaDeletionStarted =>
+        validReplicas.foreach { replica =>
+          replicaState.put(replica, ReplicaDeletionStarted)
+          controllerBrokerRequestBatch.addStopReplicaRequestForBrokers(Seq(replicaId),
+            replica.topic,
+            replica.partition,
+            deletePartition = true,
+            callbacks(replica.topicAndPartition).stopReplicaResponseCallback)
+        }
+      case ReplicaDeletionIneligible =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionIneligible))
+      case ReplicaDeletionSuccessful =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionSuccessful))
+      case NonExistentReplica =>
+        validReplicas.foreach { replica =>
+          val currentAssignedReplicas = controllerContext.partitionReplicaAssignment(replica.topicAndPartition)
+          controllerContext.partitionReplicaAssignment.put(replica.topicAndPartition, currentAssignedReplicas.filterNot(_ == replica.replica))
+          replicaState.remove(replica)
+        }
+    }
+  }
+  private def removeReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    var results = Map.empty[TopicAndPartition, LeaderIsrAndControllerEpoch]
+    var remaining = partitions
+    while (remaining.nonEmpty) {
+      val (successfulRemovals, removalsToRetry) = doRemoveReplicasFromIsr(replicaId, remaining)
+      results ++= successfulRemovals
+      remaining = removalsToRetry
+    }
+    results
+  }
+
+  private def doRemoveReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    val (currentLeaderAndIsrs, candidateLeaderAndIsrs, partitionsWithNoLeaderAndIsrInZk, failedStateReads) = getTopicPartitionStatesFromZk(replicaId, partitions)
+    val partitionsWhoseIsrExcludesReplica = currentLeaderAndIsrs -- candidateLeaderAndIsrs.keys
+    val (partitionsWithEmptyIsr, partitionsWithNonEmptyIsr) = candidateLeaderAndIsrs.partition { case (partition, leaderAndIsr) => leaderAndIsr.isr.isEmpty }
```

Comment:
There was one typo in getTopicPartitionStatesFromZk:
`val candidateIsr = leaderIsrAndControllerEpoch.leaderAndIsr.isr.filterNot(_ != replicaId)`
should be:
`val candidateIsr = leaderIsrAndControllerEpoch.leaderAndIsr.isr.filter(_ != replicaId)`

but otherwise we actually do the topic config checks only on the partitions whose isr is empty after removal. Note that the `candidateLeaderAndIsrs` returned is used to find `partitionsWithEmptyIsr` which is what we pass into `leaderAndIsrBasedOnLogConfigs`.